src/ru/nsu/ccfit/zuev/osu/MainScene.java

```diff
@@ -87,8 +88,10 @@ public class MainScene implements IUpdateHandler {
     private final float[] peakLevel = new float[120];
     private final float[] peakDownRate = new float[120];
     private final float[] peakAlpha = new float[120];
-    private List<TimingPoint> timingPoints;
-    private TimingPoint currentTimingPoint, lastTimingPoint, firstTimingPoint;
+    private LinkedList<TimingControlPoint> timingControlPoints;
+    private LinkedList<EffectControlPoint> effectControlPoints;
+    private TimingControlPoint currentTimingPoint, lastTimingPoint, firstTimingPoint;
+    private EffectControlPoint currentEffectPoint;
```

Comment:
Can't you do the same you did in `GameScene` here pointing one reference per control point type?

Reply:
I suppose `firstTimingPoint` can be removed, but `lastTimingPoint` is used to calculate the beat offset of the parallax effect:

https://github.com/osudroid/osu-droid/blob/d16bcad3924c7e82fe8858a74991ebd0ff1fb522/src/ru/nsu/ccfit/zuev/osu/MainScene.java#L557-L559

https://github.com/osudroid/osu-droid/blob/d16bcad3924c7e82fe8858a74991ebd0ff1fb522/src/ru/nsu/ccfit/zuev/osu/MainScene.java#L566-L569

https://github.com/osudroid/osu-droid/blob/d16bcad3924c7e82fe8858a74991ebd0ff1fb522/src/ru/nsu/ccfit/zuev/osu/MainScene.java#L608-L611

Reply:
Have applied this in https://github.com/osudroid/osu-droid/pull/384/commits/f7959825f7fd680e80aec0524d6ff68ef889b7b3.